index.js

```diff
@@ -164,6 +164,21 @@ class MetasysServerApi {
     this.requestOriginal = requestObject || request;
   }
 
+  scheduleRefresh(timeout) {
+    const self = this;
+    const interval = timeout || 1500000; // 25 minutes
```

Comment:
You may want to consider taking this down to under 1 minute, say 55 seconds. The reason is that the session length in SMP is configurable and can be 1 minute at a minimum. If you go less than a minute then you are guaranteed that the configuration for the user can't be less than the interval you've defined. I know this seems frequent, but the MUI client generally refreshes every 30 seconds or so as long as the user is interacting with the application.

Another option would be to get the expiration time from the token here without a timeout being passed in and instead have the access token passed in. You could then accomplish this with something like:

```js
scheduleRefresh(accessToken) {
    const self = this;
    async function refreshToken() {
        let accessToken = (await self.get('/refreshToken')).accessToken;
        self.updateAccessToken(accessToken);
        self.scheduleRefresh(accessToken);
    }
    setTimeout(refreshToken, getRefreshTimeout(accessToken));
}

getRefreshTimeout(accessToken) {
    const refreshBeforeExpiration = 60000; // 1 minute
    return Math.max(1000, getTokenExpiration(accessToken) - refreshBeforeExpiration);
}

getTokenExpiresIn(accessToken) {
    const pieces = accessToken.split('.');
    const claims = JSON.parse(atob(pieces[1]);
 
    // The exp claims is a unix timestamp in seconds that needs to be converted to ms
    return (claims['exp'] * 1000) - Date.now();    
}
```
With something close to this you could call `scheduleRefresh(accessToken)` from the login method and all of the interval management will be handled based on the information in the token itself. The refresh will be attempted 1 minute before expiration and not any more often than once a second.

Reply:
@musicfuel I was thinking that the consumer could set it.

Also, I was thinking I should just set the interval based on the `expires` field rather than guessing.

Reply:
> getTokenExpiresIn(accessToken) { const pieces = accessToken.split('.'); const claims = JSON.parse(atob(pieces[1]); // The exp claims is a unix timestamp in seconds that needs to be converted to ms return (claims['exp'] * 1000) - Date.now(); }

now way! Dude, you gave me an `expires` field.

Reply:
Also, I don't think we should take advantage of undocumented information.

Reply:
Sure. The `exp` claim isn't really undocumented as it is part of the JWT standard, but that is true. I completely forgot about the fact that the endpoint is providing the expiration time. If that were passed to the `scheduleRefresh` you would be able to basically follow the same pattern, but much more cleanly.

Reply:
Yep. And I decided that instead of using `Date.now()` to use `response.headers.date` so that I could get server time. I had noticed that my timeouts were awfully short. that was because the server time was 15 seconds different than mine.

Reply:
I'm testing with a user whose expiration is 1 minute.